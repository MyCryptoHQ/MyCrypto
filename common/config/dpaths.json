{
  "TREZOR": [
    {
      "label": "TREZOR (ETH)",
      "value": "m/44'/60'/0'/0"
    },
    {
      "label": "TREZOR (ETC)",
      "value": "m/44'/61'/0'/0"
    },
    {
      "label": "Testnet",
      "value": "m/44'/1'/0'/0"
    }
  ]
}
