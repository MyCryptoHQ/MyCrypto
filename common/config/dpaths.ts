const ETH_DEFAULT = {
  label: 'Default (ETH)',
  value: "m/44'/60'/0'/0"
};

const ETH_TREZOR = {
  label: 'TREZOR (ETH)',
  value: "m/44'/60'/0'/0"
};

const ETH_LEDGER = {
  label: 'Ledger (ETH)',
  value: "m/44'/60'/0'"
};

const ETC_LEDGER = {
  label: 'Ledger (ETC)',
  value: "m/44'/60'/160720'/0'"
};

const ETC_TREZOR = {
  label: 'TREZOR (ETC)',
  value: "m/44'/61'/0'/0"
};

const TESTNET = {
  label: 'Testnet',
  value: "m/44'/1'/0'/0"
};

const EXPANSE = {
  label: 'Expanse',
  value: "m/44'/40'/0'/0"
};

const TREZOR = [ETH_TREZOR, ETC_TREZOR, TESTNET];

const MNEMONIC = [ETH_DEFAULT, ETH_LEDGER, ETC_LEDGER, ETC_TREZOR, TESTNET, EXPANSE];

const LEDGER = [ETH_LEDGER, ETC_LEDGER, TESTNET];

export default {
  TREZOR,
  MNEMONIC,
  LEDGER
};
