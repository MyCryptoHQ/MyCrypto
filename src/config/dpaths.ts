import { DPath } from '@types';

const ETH_DEFAULT: DPath = {
  label: 'Default (ETH)',
  value: "m/44'/60'/0'/0"
};

const ETH_TREZOR: DPath = {
  label: 'TREZOR (ETH)',
  value: "m/44'/60'/0'/0"
};

const ETH_LEDGER: DPath = {
  label: 'Ledger (ETH)',
  value: "m/44'/60'/0'"
};

const ETH_LEDGER_LIVE: DPath = {
  label: 'Ledger Live (ETH)',
  value: "m/44'/60'/0'/0/0",
  isHardened: true,
  getIndex: (addressIndex): string => `m/44'/60'/${addressIndex}'/0/0`
};

const ETC_LEDGER: DPath = {
  label: 'Ledger (ETC)',
  value: "m/44'/60'/160720'/0'"
};

const ETC_TREZOR: DPath = {
  label: 'TREZOR (ETC)',
  value: "m/44'/61'/0'/0"
};

const ETH_TESTNET: DPath = {
  label: 'Testnet (ETH)',
  value: "m/44'/1'/0'/0"
};

const EXP_DEFAULT: DPath = {
  label: 'Default (EXP)',
  value: "m/44'/40'/0'/0"
};

const UBQ_DEFAULT: DPath = {
  label: 'Default (UBQ)',
  value: "m/44'/108'/0'/0"
};

const POA_DEFAULT: DPath = {
  label: 'Default (POA)',
  value: "m/44'/60'/0'/0"
};

const TOMO_DEFAULT: DPath = {
  label: 'Default (TOMO)',
  value: "m/44'/889'/0'/0"
};

const MUSIC_DEFAULT: DPath = {
  label: 'Default (MUSIC)',
  value: "m/44'/184'/0'/0"
};

const EGEM_DEFAULT: DPath = {
  label: 'Default (EGEM)',
  value: "m/44'/1987'/0'/0"
};

const CLO_DEFAULT: DPath = {
  label: 'Default (CLO)',
  value: "m/44'/820'/0'/0"
};

const ETH_SINGULAR: DPath = {
  label: 'SingularDTV',
  value: "m/0'/0'/0'"
};

const RSK_TESTNET: DPath = {
  label: 'Testnet (RSK)',
  value: "m/44'/37310'/0'/0"
};

const RSK_MAINNET: DPath = {
  label: 'Mainnet (RSK)',
  value: "m/44'/137'/0'/0"
};

const GO_DEFAULT: DPath = {
  label: 'Default (GO)',
  value: "m/44'/6060'/0'/0"
};

const ATH_DEFAULT: DPath = {
  label: 'Default (ATH)',
  value: "m/44'/1620'/0'/0"
};

const ETHO_DEFAULT: DPath = {
  label: 'Default (ETHO)',
  value: "m/44'/1313114'/0'/0"
};

const MIX_DEFAULT: DPath = {
  label: 'Default (MIX)',
  value: "m/44'/76'/0'/0"
};

const REOSC_DEFAULT: DPath = {
  label: 'Default (REOSC)',
  value: "m/44'/2894'/0'/0"
};

const ARTIS_SIGMA1: DPath = {
  label: 'Sigma1 (ATS)',
  value: "m/44'/60'/0'/0"
};

const ARTIS_TAU1: DPath = {
  label: 'Tau1 (ATS)',
  value: "m/44'/60'/0'/0"
};

const THUNDERCORE_DEFAULT: DPath = {
  label: 'Default (THUNDERCORE)',
  value: "m/44'/1001'/0'/0"
};

const ETI_DEFAULT: DPath = {
  label: 'Default (ETI)',
  value: "m/44'/464'/0'/0"
};
const WEB_DEFAULT: DPath = {
  label: 'Default (WEB)',
  value: "m/44'/227'/0'/0"
};

const METADIUM_DEFAULT: DPath = {
  label: 'Default (METADIUM)',
  value: "m/44'/916'/0'/0"
};

const DEXON_DEFAULT: DPath = {
  label: 'Default (DEXON)',
  value: "m/44'/237'/0'/0"
};

const ASK_DEFAULT: DPath = {
  label: 'Default (ASK)',
  value: "m/44'/2221'/0'/0"
};

const ASK_TREZOR: DPath = {
  label: 'TREZOR (ASK)',
  value: "m/44'/2221'/0'/0"
};

const AUX_DEFAULT: DPath = {
  label: 'Default (AUX)',
  value: "m/44'/344'/0'/0"
};

const ERE_DEFAULT: DPath = {
  label: 'Default (ERE)',
  value: "m/44'/466'/0'/0"
};

const VOLTA_DEFAULT: DPath = {
  label: 'Default (VOLTA)',
  value: "m/44'/73799'/0'/0"
};

const EWC_DEFAULT: DPath = {
  label: 'Default (EWC)',
  value: "m/44'/246'/0'/0"
};

const AVAX_DEFAULT: DPath = {
  label: 'Default (AVAX)',
  value: "m/44'/60'/0'/0"
};

export const DPathsList = {
  ETH_DEFAULT,
  ETH_TREZOR,
  ETH_LEDGER,
  ETH_LEDGER_LIVE,
  ETC_LEDGER,
  ETC_TREZOR,
  ETH_TESTNET,
  EXP_DEFAULT,
  UBQ_DEFAULT,
  POA_DEFAULT,
  TOMO_DEFAULT,
  MUSIC_DEFAULT,
  EGEM_DEFAULT,
  CLO_DEFAULT,
  RSK_MAINNET,
  RSK_TESTNET,
  GO_DEFAULT,
  ATH_DEFAULT,
  ETHO_DEFAULT,
  MIX_DEFAULT,
  REOSC_DEFAULT,
  ARTIS_SIGMA1,
  ARTIS_TAU1,
  THUNDERCORE_DEFAULT,
  ETI_DEFAULT,
  WEB_DEFAULT,
  METADIUM_DEFAULT,
  DEXON_DEFAULT,
  ASK_DEFAULT,
  ASK_TREZOR,
  AUX_DEFAULT,
  ETH_SINGULAR,
  ERE_DEFAULT,
  VOLTA_DEFAULT,
  EWC_DEFAULT,
  AVAX_DEFAULT
};

// PATHS TO BE INCLUDED REGARDLESS OF WALLET FORMAT
// const EXTRA_PATHS = [ETH_SINGULAR];

// Full length deterministic wallet paths from BIP44
// https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki
// normal path length is 4, ledger is the exception at 3

// m / purpose' / coin_type' / account' / change / address_index
//   |          |            |          |        |
//   | constant |   index    |  index   | 0 or 1 |
//   |__________|____________|__________|________|

// whitespace strings are evaluated the same way as nospace strings, except they allow optional spaces between each portion of the string
// ie. "m / 44' / 0' / 0'" is valid, "m / 4 4' / 0' / 0'" is invalid
export const dPathRegex = /m\/4[4,9]'\/[0-9]+'\/[0-9]+('+$|'+(\/[0-1]+$))/;
// const whitespaceDPathRegex = /m\s*\/\s*44'\s*\/\s*[0-9]+\'\s*\/\s*[0-9]+(\'+$|\'+\s*(\/\s*[0-1]+$))/;

/**
 * Due to limitations in the Ledger ETH application, only derivation paths starting with
 * `m/44'/60'` and `m/44'/1'` can be checked.
 */
export const LEDGER_DERIVATION_PATHS: DPath[] = [
  ETH_DEFAULT,
  ETH_LEDGER,
  ETC_LEDGER,
  ETH_TESTNET,
  ETH_LEDGER_LIVE
];

/**
 * While Trezor does support hardened paths, it'd be very tedious for the user to check all the
 * paths currently, since the user would have to confirm each address individually.
 */
export const TREZOR_DERIVATION_PATHS: DPath[] = [
  ...Object.values(DPathsList).filter((path) => !path.isHardened)
];
