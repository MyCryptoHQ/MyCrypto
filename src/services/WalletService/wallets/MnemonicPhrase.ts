import { isValidMnemonic, HDNode, fromMnemonic } from 'ethers/utils/hdnode';

import Wallet from './Wallet';

import { WalletId } from '@types';

import { MnemonicPhraseResult } from './types';
import { DPathsList } from '@config/dpaths';
import { getFullPath } from './helpers';

export default class MnemonicPhrase implements Wallet {
  private readonly mnemonicPhrase: string;
  private readonly password: string;

  public constructor(mnemonicPhrase: string, password: string) {
    this.mnemonicPhrase = mnemonicPhrase;
    this.password = password;
  }

  public async getAddress(dPath: DPath, index: number): Promise<MnemonicPhraseResult[]> {
    const hdNode = await this.getHDNode();
    const results: MnemonicPhraseResult[] = [];

    const path = getFullPath(dPath, index);

    results.push({
      type: WalletId.MNEMONIC_PHRASE_NEW,
      address: hdNode.derivePath(path).address,
      withPassword: !!this.password,
      path
    });

    // Extra check without a password if a password was specified
    if (this.password) {
      const passwordlessHdNode = await this.getHDNode(false);

      results.push({
        type: WalletId.MNEMONIC_PHRASE_NEW,
        address: passwordlessHdNode.derivePath(path).address,
        withPassword: false,
        path
      });
    }

    return results;
  }

  public async initialize(): Promise<void> {
    if (!isValidMnemonic(this.mnemonicPhrase)) {
      throw new Error('The mnemonic phrase you provided is invalid.');
    }
  }

  public getDPaths(): DPath[] {
    return Object.values(DPathsList);
  }

  /**
   * Get an instance of the HDNode class.
   *
   * @param {boolean} withPassword Whether to get the HDNode initialized with a password or not.
   * @return {Promise<HDNode>} A Promise with an instance of the HDKey class.
   */
  private async getHDNode(withPassword: boolean = true): Promise<HDNode> {
    return fromMnemonic(this.mnemonicPhrase, undefined, withPassword ? this.password : undefined);
  }
}
