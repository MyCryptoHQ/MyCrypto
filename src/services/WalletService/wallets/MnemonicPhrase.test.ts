import { DPathsList } from '@config/dpaths';

import MnemonicPhrase from './MnemonicPhrase';

const VALID_MNEMONIC = 'measure awake inject because stay profit soup dawn rare wave news cook';
const INVALID_MNEMONIC = 'setup safe obscure bus poem very fatal sock color design bridge garden';

describe('MnemonicPhrase', () => {
  it('generates an address from a derivation path', async () => {
    const wallet = new MnemonicPhrase(VALID_MNEMONIC, '');
    await expect(wallet.getAddress(DPathsList.ETH_DEFAULT, 0)).resolves.toMatchSnapshot();
    await expect(wallet.getAddress(DPathsList.ETH_DEFAULT, 1)).resolves.toMatchSnapshot();
  });

  it('generates an address from a hardened derivation path', async () => {
    const wallet = new MnemonicPhrase(VALID_MNEMONIC, '');
    await expect(wallet.getAddress(DPathsList.ETH_LEDGER_LIVE, 3)).resolves.toMatchSnapshot();
    await expect(wallet.getAddress(DPathsList.ETH_LEDGER_LIVE, 4)).resolves.toMatchSnapshot();
  });

  it('generates different addresses with a passphrase provided', async () => {
    const wallet = new MnemonicPhrase(VALID_MNEMONIC, 'foobar');
    await expect(wallet.getAddress(DPathsList.ETH_DEFAULT, 0)).resolves.toMatchSnapshot();
    await expect(wallet.getAddress(DPathsList.ETH_DEFAULT, 1)).resolves.toMatchSnapshot();
  });

  it('throws an error on invalid mnemonic phrases', async () => {
    const wallet = new MnemonicPhrase(INVALID_MNEMONIC, '');
    await expect(wallet.initialize()).rejects.toThrow();
  });
});
