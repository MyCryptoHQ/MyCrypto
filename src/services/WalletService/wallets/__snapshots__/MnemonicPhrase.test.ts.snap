// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MnemonicPhrase generates an address from a derivation path 1`] = `
Array [
  Object {
    "address": "0xBa0310bEE9fDd582530cd1cD0C29aCF7f03cC548",
    "path": "m/44'/60'/0'/0/0",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": false,
  },
]
`;

exports[`MnemonicPhrase generates an address from a derivation path 2`] = `
Array [
  Object {
    "address": "0x3B2006f42edfb0ABFDBB848A7fBe028A8E984e65",
    "path": "m/44'/60'/0'/0/1",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": false,
  },
]
`;

exports[`MnemonicPhrase generates an address from a hardened derivation path 1`] = `
Array [
  Object {
    "address": "0x9F8708a7091B34797a6810ab01Eacf2594e1e93d",
    "path": "m/44'/60'/3'/0/0",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": false,
  },
]
`;

exports[`MnemonicPhrase generates an address from a hardened derivation path 2`] = `
Array [
  Object {
    "address": "0x4029471cEA2B1cB785A5045d517523E91AB92569",
    "path": "m/44'/60'/4'/0/0",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": false,
  },
]
`;

exports[`MnemonicPhrase generates different addresses with a passphrase provided 1`] = `
Array [
  Object {
    "address": "0x91F81204910ab591d8ab1e7ab5db42D8FB48c646",
    "path": "m/44'/60'/0'/0/0",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": true,
  },
  Object {
    "address": "0xBa0310bEE9fDd582530cd1cD0C29aCF7f03cC548",
    "path": "m/44'/60'/0'/0/0",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": false,
  },
]
`;

exports[`MnemonicPhrase generates different addresses with a passphrase provided 2`] = `
Array [
  Object {
    "address": "0x65CF9d2AC5DFe4A27d22A48A9b898d7B01Ee1c4a",
    "path": "m/44'/60'/0'/0/1",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": true,
  },
  Object {
    "address": "0x3B2006f42edfb0ABFDBB848A7fBe028A8E984e65",
    "path": "m/44'/60'/0'/0/1",
    "type": "MNEMONIC_PHRASE_NEW",
    "withPassword": false,
  },
]
`;
